mlkem/reduce.c

```diff
@@ -9,14 +9,40 @@
  * Description: Montgomery reduction; given a 32-bit integer a, computes
  *              16-bit integer congruent to a * R^-1 mod q, where R=2^16
  *
- * Arguments:   - int32_t a: input integer to be reduced;
- *                           has to be in {-q2^15,...,q2^15-1}
+ * Arguments:   - int32_t a: input integer to be reduced
  *
- * Returns:     integer in {-q+1,...,q-1} congruent to a * R^-1 modulo q.
+ * Returns:     integer congruent to a * R^-1 modulo q
+ *
+ *              Bounds: If |a| < q * C, then the return value
+ *              has absolute value < q (C/2^16 + 1/2).
+ *
+ *              Notable special cases:
+ *              - The Montgomery multiplication of a value of absolute value
+ *                < q * C with a signed-canonical value ( < q/2 ) has
+ *                absolute value q * (0.0254 * C + 1/2).
+ *              - The Montgomery multiplication of a value of absolute value
+ *                < q * C with a value t of |t| < q has absolute value
+ *                < q * (0.0508 * C + 1/2).
+ *              - The Montgomery multiplication of a value of absolute value
+ *                < C with a value of abs < q has absolute value
+ *                < q (C/2^16 + 1/2).
  **************************************************/
 int16_t montgomery_reduce(int32_t a) {
   int16_t t;
 
+  // Bounds on paper
+  //
+  // - Case |a| < q * C
+  //
+  // |t| <= |a|/2^16 + |t|*q/2^16
+  //      < q * C / 2^16 + q/2
+  //      = q (C/2^16 + 1/2)
+  //
+  // - Case |a| < (q/2) * C * q
+  //
+  // Replace C -> C * q in the above and estimate
+  // q / 2^17 < 0.0254.
+
```

Comment:
Can you add two BOUND() macros here to check the pre-condition and the bound of the result?

Reply:
No, because everything is parametric, we don't know what `C` is. But that's what `test_bounds.py` is for.

Reply:
OK